nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+from pytorch_lightning import Trainer
+from pytorch_lightning.loggers import TensorBoardLogger, WandbLogger
+from torch import Tensor, nn
+from torch.autograd import grad as torch_grad
+from torch.utils.tensorboard.writer import SummaryWriter
+
+import nemo
+from nemo.collections.tts.models import FastPitchModel
+from nemo.collections.tts.models.base import SpectrogramGenerator
+from nemo.collections.tts.modules.spectrogram_enhancer import Discriminator, Generator, mask
+from nemo.core import Exportable, ModelPT
+from nemo.utils import logging
+
+
+def type_as_recursive(e, source: Tensor):
+    if isinstance(e, (list, tuple)):
+        return [type_as_recursive(elem, source) for elem in e]
+    elif isinstance(e, dict):
+        return {key: type_as_recursive(value, source) for key, value in e.items()}
+    elif isinstance(e, Tensor):
+        return e.type_as(source)
+    else:
+        return e
+
+
+def gradient_penalty(images, output, weight=10):
+    batch_size = images.shape[0]
+    gradients = torch_grad(
+        outputs=output,
+        inputs=images,
+        grad_outputs=torch.ones(output.size(), device=images.device),
+        create_graph=True,
+        retain_graph=True,
+        only_inputs=True,
+    )[0]
+
+    gradients = gradients.reshape(batch_size, -1)
+    return weight * ((gradients.norm(2, dim=1) - 1) ** 2).mean()
+
+
+def calc_pl_lengths(styles, images):
+    device = images.device
+    num_pixels = images.shape[2] * images.shape[3]
+    pl_noise = torch.randn(images.shape, device=device) / math.sqrt(num_pixels)
+    outputs = (images * pl_noise).sum()
+
+    pl_grads = torch_grad(
+        outputs=outputs,
+        inputs=styles,
+        grad_outputs=torch.ones(outputs.shape, device=device),
+        create_graph=True,
+        retain_graph=True,
+        only_inputs=True,
+    )[0]
+
+    return (pl_grads ** 2).sum(dim=2).mean(dim=1).sqrt()
+
+
+def gen_hinge_loss(fake):
+    return fake.mean()
+
+
+def hinge_loss(real, fake):
+    return (F.relu(1 + real) + F.relu(1 - fake)).mean()
+
+
+def consistency_loss(condition, output, lengths):
+    *_, w, h = condition.shape
+    w, h = w // 4, h
+
+    condition = torch.nn.functional.interpolate(condition, size=(w, h), mode="bilinear", antialias=True)
+    output = torch.nn.functional.interpolate(output, size=(w, h), mode="bilinear", antialias=True)
+
+    dist = (condition - output).abs()
+    dist = mask(dist, lengths)
+    return (dist / rearrange(lengths, "b 1 -> b 1 1 1")).sum(dim=-1).mean()
+
+
+class SpectrogramEnhancerModel(ModelPT, Exportable):
+    """GAN-based model to add details to blurry spectrograms from TTS models like Tacotron or FastPitch."""
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None) -> None:
+        self.spectrogram_model = None
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.init_spectrogram_model()
+
+        self.G = Generator(
+            n_bands=cfg["n_bands"],
```

Comment:
try to maintain uniformity across all models by using `cfg.n_bands` instead of `cfg["n_bands"]`

Reply:
Done